### ISSUE TITLE
`clang++ -E -dM -` runs in C mode (fails with either `/etc/clang/clang++.cfg` or an explicit `-std=c++NN` argument)

### ISSUE THREAD

**Issue by thesamesam** [llvm/llvm-project#61641]
The driver code for preprocessing currently forces C mode when stdin is used as an argument, even when the used driver is `clang++` (or otherwise originally in C++ mode).

When invoked as `clang++`, `clang++` reads `/etc/clang/clang++.cfg` accordingly where C++ specific parameters might be passed.

The special handling has been in Clang since 1688f1a7e0add08f7799485e9eb41a7ed440391d and still exists in the same form now:
https://github.com/llvm/llvm-project/blob/89a1af749166627704cdf9d676455d32616c2c06/clang/lib/Driver/Driver.cpp#L2606

This originally came up at https://reviews.llvm.org/D133375#4214214.

Steps to reproduce (method 1, config file):
```shell
mkdir -p /etc/clang
echo "-std=c++17" >> /etc/clang/clang++.cfg
clang++ -E -dM -
```

Steps to reproduce (method 2, explicit argument):
```shell
echo "int main();" | clang++ -std=c++17  -E -dM -
```

Passing `-x c++` is a workaround.

This is problematic for e.g. Meson which tries to invoke the active C++ compiler to determine any predefined macros. This is not specific to configuration files, but is worse with it. This issue didn't occur when using the legacy `CLANG_DEFAULT_STD_CXX` CMake option though.

cc @tambry @MaskRay @mgorny @eli-schwartz

**Comment by eli-schwartz** [comment]
> If Meson's intention is to check C++ macros (e.g. `__cplusplus`), I think it should specify `-x c++` or refrain from using `-` as input.

Checking C++ macros is best done by source code, all Meson wants to do is stuff like "get the gcc/lcc version from `__GNUC_MINOR__` / `__LCC_MINOR__` and stuff" or "check whether `__i386__` is defined when the build machine is 64-bit and declare that the current build machine is actually 32-bit also".



> The difference is that we report an error for the `-std=c++17` diagnostic. The diagnostic as an error is a long-term behavior, so I am unsure we should change it now.
> (The current behavior is that the loaded file respects `-x` and implicit C=>C++ conversion, e.g. `clang -xc++` loads `clang++.cfg`.)

I am slightly curious why a global compiler-name.cfg file loaded by default needs to set the std to begin with, but I'm also curious why clang reports errors for information that comes from a cfg file rather than the command line. And I'm also curious, if the loaded file respects -x and implicit C=>C++ conversion... why doesn't it respect implicit C++=>C conversion? If the detected mode for reading stdin is to switch to C, shouldn't it process the input file as stdin *first*, and then not load clang++.cfg, only clang.cfg?

**Comment by MaskRay** [comment]
In `std::unique_ptr<Compilation> C(TheDriver.BuildCompilation(Args));`, `loadDefaultConfigFiles(ExpCtx)` is called before `BuildInputs(C->getDefaultToolChain(), *TranslatedArgs, Inputs);`. `loadDefaultConfigFiles(ExpCtx)` uses the executable name (`clang++`) to decide whether `clang++.cfg` or `clang.cfg` should be loaded by default.

Then, Clang Driver processes `BuildInputs(C->getDefaultToolChain(), *TranslatedArgs, Inputs);`. It inherits the strange GCC rule that `clang++ -E -dM -` runs in the C mode.
The default option `-std=c++*` leads to an error in the C mode.

So I don't think this is strange. Whether `clang++.cfg` or `clang.cfg` is read is not dependent on the input file.
Clang Driver may process multiple input files in one invocation. If different `.cfg` files just because input files are of different modes, the behavior would be weirder.

**Comment by eli-schwartz** [comment]
> So I don't think this is strange. Whether `clang++.cfg` or `clang.cfg` is read is not dependent on the input file.

I didn't say it was dependent on that. However, *you* said that it's dependent on -xc++, which seems equally weird to me, so I was wondering... why one and not the other?

> Clang Driver may process multiple input files in one invocation. If different `.cfg` files just because input files are of different modes, the behavior would be weirder.

But -x is documented to only apply to files listed up until the next -x option, so, just like you can have multiple input files of different modes, so too can you have multiple -x options of different modes.

What does clang do here? Does it read all .cfg files? Does it only read the one corresponding to the first -x option?

Whichever one is the case, why is it less weird for -x than for the input file itself?

**Comment by MaskRay** [comment]
> > So I don't think this is strange. Whether `clang++.cfg` or `clang.cfg` is read is not dependent on the input file.
> 
> I didn't say it was dependent on that. However, _you_ said that it's dependent on -xc++, which seems equally weird to me, so I was wondering... why one and not the other?
> 
> This is why my question was directly a response to your statement about respecting -x.

The following sentence from a previous reply of mine is wrong and should be ignored. Sorry for the confusion:

"(The current behavior is that the loaded file respects -x and implicit C=>C++ conversion, e.g. clang -xc++ loads clang++.cfg.)"

> > Clang Driver may process multiple input files in one invocation. If different `.cfg` files just because input files are of different modes, the behavior would be weirder.
> 
> But -x is documented to only apply to files listed up until the next -x option, so, just like you can have multiple input files of different modes, so too can you have multiple -x options of different modes.
> 
> What does clang do here? Does it read all .cfg files? Does it only read the one corresponding to the first -x option?
> 
> Whichever one is the case, why is it less weird for -x than for the input file itself?

Clang reads either `clang.cfg` or `clang++.cfg`, not both. The file is dependent on input file modes.

**Comment by eli-schwartz** [comment]
Ah okay that makes more sense. :D Thanks for clarifying.

**Comment by eli-schwartz** [comment]
> The difference is that we report an error for the` -std=c++17` diagnostic. The diagnostic as an error is a long-term behavior, so I am unsure we should change it now.

It still feels a bit to me like it would be a beneficial change for clang to recognize when a diagnostic is being issued solely due to options that were parsed from a cfg file, and either suppress that diagnostic or make it non-fatal.

GCC spec files are deep sorcery, but I believe it manages to avoid this kind of issue by acting as a programmable hook into the machinery, so you can e.g. actually adapt to what's being done. Feels a bit more comfy than spec files for that reason alone...

**Comment by tambry** [comment]
`-x c++` in `clang++.cfg` isn't a workaround unfortunately.
It breaks invocations of `clang -o executable object.o`, i.e. basically nothing works.

Example:
```
$ echo 'int main() {}' > test.cpp
$ clang++ test.cpp -o test.o
$ clang++ -o test test.o
test.o:1:1: error: expected unqualified-id
<U+007F>ELF<U+0002><U+0001><U+0001><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0003><U+0000>><U+0000><U+0001><U+0000><U+0000><U+0000>`<U+0016><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000>@<U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><B0><U+0016><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000><U+0000> [truncated, there's lots of this]
```
The only viable option remains reverting 3baadff896eddfa09fcd458d120bac95491992a7.

**Comment by tambry** [comment]
https://github.com/mesonbuild/meson/commit/4ad792e158b6059eb847dd0562aff9bd7029981f essentially fixes the Meson issue, but due to a small bug https://github.com/mesonbuild/meson/pull/13731 is required.  
Instead of reverting 3baadff896eddfa09fcd458d120bac95491992a7 I've moved to using resource files since Meson is the only real-world breakage I've encountered and it'll be fixed in the next release.

As I understand from previous discussion we want to:
1. Match GCC's behaviour.
2. Keep issuing an error for the mismatch unlike GCC.

I'll close this out and hope any new ecosystem code will come to cope with this edgecase. :slightly_smiling_face:
